File: backend/services/recruiterAgentService.js
/**
 * Recruiter AI Agent Service
 *
 * Reads all previously generated reports for a candidate,
 * sends a consolidated prompt to GROQ, and saves one
 * RecruiterRecommendation document to MongoDB.
 *
 * No existing reports are regenerated or modified.
 */

const mongoose = require('mongoose');
const ResumeReport         = require('../models/ResumeReport');
const CodingReport         = require('../models/CodingReport');
const TechnicalInterviewReport = require('../models/TechnicalInterviewReport');
const HRInterviewReport    = require('../models/HRInterviewReport');
const PlacementRecommendation  = require('../models/PlacementRecommendation');
const RecruiterRecommendation  = require('../models/RecruiterRecommendation');

// ── helpers ───────────────────────────────────────────────────────────────────

const safeAvg = (docs, field) => {
  const vals = docs.map(d => d[field]).filter(v => typeof v === 'number' && isFinite(v));
  return vals.length ? Math.round((vals.reduce((s, v) => s + v, 0) / vals.length) * 10) / 10 : null;
};

/**
 * Collect the most recent document of each report type for a given userId.
 * Returns a plain object with all data needed to build the GROQ prompt.
 */
const collectCandidateData = async (userId) => {
  const uid = new mongoose.Types.ObjectId(userId);

  const [
    resumeReports,
    codingReports,
    techReports,
    hrReports,
    placementRecs
  ] = await Promise.all([
    ResumeReport.find({ userId: uid }).sort({ createdAt: -1 }).limit(1).lean(),
    CodingReport.find({ userId: uid }).sort({ createdAt: -1 }).limit(5).lean(),
    TechnicalInterviewReport.find({ userId: uid }).sort({ createdAt: -1 }).limit(20).lean(),
    HRInterviewReport.find({ userId: uid }).sort({ createdAt: -1 }).limit(1).lean(),
    PlacementRecommendation.find({ userId: uid }).sort({ createdAt: -1 }).limit(1).lean()
  ]);

  const resume    = resumeReports[0]    || null;
  const hr        = hrReports[0]        || null;
  const placement = placementRecs[0]    || null;

  const codingAvg  = safeAvg(codingReports, 'score');
  const techAvg    = safeAvg(techReports, 'overallScore');
  const techCommAvg = safeAvg(techReports, 'communicationScore');

  return {
    resume,
    codingReports,
    techReports,
    hr,
    placement,
    // Derived averages
    codingAvg,
    techAvg,
    techCommAvg,
    // Source IDs for the recommendation document
    resumeReportId:           resume?._id || null,
    codingReportId:           codingReports[0]?._id || null,
    technicalInterviewId:     techReports[0]?._id || null,
    hrInterviewId:            hr?._id || null,
    placementRecommendationId: placement?._id || null,
  };
};

/**
 * Build a concise but information-rich text prompt from the collected data.
 */
const buildPrompt = (data) => {
  const { resume, codingAvg, techAvg, techCommAvg, hr, placement } = data;

  const lines = [];

  if (resume) {
    lines.push(`ATS Score: ${resume.atsScore ?? 'N/A'}/100`);
    lines.push(`Resume Score: ${resume.resumeScore ?? 'N/A'}/100`);
    lines.push(`Recommended Role: ${resume.recommendedRole || 'N/A'}`);
    lines.push(`Skills: ${(resume.skills || []).slice(0, 10).join(', ') || 'N/A'}`);
    lines.push(`Resume Strengths: ${(resume.strengths || []).slice(0, 3).join('; ') || 'N/A'}`);
    lines.push(`Resume Weaknesses: ${(resume.weaknesses || []).slice(0, 3).join('; ') || 'N/A'}`);
  }

  if (codingAvg !== null) {
    lines.push(`Coding Avg Score: ${codingAvg}/10`);
    const verdicts = data.codingReports.map(r => r.verdict).filter(Boolean);
    if (verdicts.length) lines.push(`Coding Verdicts: ${verdicts.join(', ')}`);
  }

  if (techAvg !== null) {
    lines.push(`Technical Interview Avg Score: ${techAvg}/10`);
    lines.push(`Communication Score (Tech): ${techCommAvg ?? 'N/A'}/10`);
    const strengths = data.techReports.map(r => r.strength).filter(Boolean).slice(0, 3);
    const weaknesses = data.techReports.map(r => r.weakness).filter(Boolean).slice(0, 3);
    if (strengths.length) lines.push(`Technical Strengths: ${strengths.join('; ')}`);
    if (weaknesses.length) lines.push(`Technical Weaknesses: ${weaknesses.join('; ')}`);
  }

  if (hr) {
    lines.push(`HR Interview Overall: ${hr.overallScore ?? 'N/A'}/10`);
    lines.push(`HR Communication: ${hr.communicationScore ?? 'N/A'}/10`);
    lines.push(`HR Confidence: ${hr.confidenceScore ?? 'N/A'}/10`);
    lines.push(`HR Professionalism: ${hr.professionalismScore ?? 'N/A'}/10`);
    lines.push(`HR Recommendation: ${hr.recommendation || 'N/A'}`);
  }

  if (placement) {
    lines.push(`Placement Readiness: ${placement.overallReadiness || 'N/A'}`);
    lines.push(`Estimated Placement Chance: ${placement.estimatedPlacementChance ?? 'N/A'}%`);
    lines.push(`Placement Recommended Role: ${placement.recommendedRole || 'N/A'}`);
    if ((placement.strengths || []).length)
      lines.push(`Placement Strengths: ${placement.strengths.slice(0, 3).join('; ')}`);
    if ((placement.skillsToImprove || []).length)
      lines.push(`Skills to Improve: ${placement.skillsToImprove.slice(0, 3).join(', ')}`);
  }

  const candidateProfile = lines.join('\n');

  return `You are a senior technical recruiter making a final hiring decision.

Below is the complete AI-generated assessment of a candidate across all evaluation modules.

--- CANDIDATE ASSESSMENT ---
${candidateProfile || 'No assessment data available yet.'}
--- END ASSESSMENT ---

Based on this data, provide your recruiter recommendation.

CRITICAL: Return ONLY a valid JSON object. No markdown. No explanation. No text before or after.
Start with { and end with }.

{
  "overallScore": <number 0-10, one decimal>,
  "recommendation": "<exactly one of: Strong Hire | Hire | Maybe | Need Another Interview | Reject>",
  "riskLevel": "<exactly one of: Low | Medium | High>",
  "technicalReadiness": <number 0-10, one decimal>,
  "communicationReadiness": <number 0-10, one decimal>,
  "cultureFit": <number 0-10, one decimal>,
  "strengths": ["<strength 1>", "<strength 2>", "<strength 3>"],
  "weaknesses": ["<weakness 1>", "<weakness 2>"],
  "summary": "<3-4 sentence final recruiter summary covering technical fit, communication, risk, and hiring recommendation>"
}`;
};

/**
 * Parse the GROQ response using multi-strategy extraction.
 * Mirrors the robust parser from fake_skill.py but in JS.
 */
const parseGroqResponse = (raw) => {
  if (!raw || !raw.trim()) return null;

  // Strategy 1: strip markdown fences and parse directly
  let cleaned = raw.trim()
    .replace(/```json?\s*/g, '')
    .replace(/```/g, '')
    .trim();
  try { return JSON.parse(cleaned); } catch (_) {}

  // Strategy 2: extract from first { to last }
  const first = raw.indexOf('{');
  const last  = raw.lastIndexOf('}');
  if (first !== -1 && last > first) {
    try { return JSON.parse(raw.slice(first, last + 1)); } catch (_) {}
  }

  return null;
};

/**
 * Main entry point — called by the controller.
 * candidateId is the userId of the candidate being evaluated.
 */
const generateRecruiterRecommendation = async (candidateId) => {
  const GROQ_API_KEY = process.env.GROQ_API_KEY;
  if (!GROQ_API_KEY) {
    throw new Error('GROQ_API_KEY not configured');
  }

  const data = await collectCandidateData(candidateId);

  const hasAnyData =
    data.resume || data.codingAvg !== null ||
    data.techAvg !== null || data.hr || data.placement;

  if (!hasAnyData) {
    throw new Error(
      'No assessment data found for this candidate. ' +
      'Complete at least one evaluation module first.'
    );
  }

  const prompt = buildPrompt(data);

  const groqRes = await fetch('https://api.groq.com/openai/v1/chat/completions', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${GROQ_API_KEY}`
    },
    body: JSON.stringify({
      model: 'llama-3.3-70b-versatile',
      messages: [
        {
          role: 'system',
          content:
            'You are a JSON-only API. Respond with ONLY a valid JSON object. ' +
            'Never include markdown, code fences, or any text outside the JSON object.'
        },
        { role: 'user', content: prompt }
      ],
      temperature: 0.2,
      max_tokens: 700
    })
  });

  if (!groqRes.ok) {
    const err = await groqRes.text();
    throw new Error(`GROQ API error: ${err}`);
  }

  const groqData  = await groqRes.json();
  const rawText   = groqData.choices[0].message.content.trim();
  const aiResult  = parseGroqResponse(rawText);

  if (!aiResult) {
    console.error('[RecruiterAgent] JSON parse failed. Raw (first 400):', rawText.slice(0, 400));
    throw new Error('Unable to parse AI response. Please try again.');
  }

  // Validate and clamp
  const VALID_RECOMMENDATIONS = ['Strong Hire', 'Hire', 'Maybe', 'Need Another Interview', 'Reject'];
  const VALID_RISKS = ['Low', 'Medium', 'High'];
  const clamp = (v, lo, hi) => Math.min(hi, Math.max(lo, Number(v) || 0));

  const recommendation = new RecruiterRecommendation({
    userId:                    candidateId,
    candidateId:               candidateId,
    resumeReportId:            data.resumeReportId,
    codingReportId:            data.codingReportId,
    technicalInterviewId:      data.technicalInterviewId,
    hrInterviewId:             data.hrInterviewId,
    placementRecommendationId: data.placementRecommendationId,

    overallScore:            Math.round(clamp(aiResult.overallScore, 0, 10) * 10) / 10,
    technicalReadiness:      Math.round(clamp(aiResult.technicalReadiness, 0, 10) * 10) / 10,
    communicationReadiness:  Math.round(clamp(aiResult.communicationReadiness, 0, 10) * 10) / 10,
    cultureFit:              Math.round(clamp(aiResult.cultureFit, 0, 10) * 10) / 10,

    recommendation: VALID_RECOMMENDATIONS.includes(aiResult.recommendation)
      ? aiResult.recommendation : 'Maybe',
    riskLevel: VALID_RISKS.includes(aiResult.riskLevel)
      ? aiResult.riskLevel : 'Medium',

    strengths:  Array.isArray(aiResult.strengths)  ? aiResult.strengths  : [],
    weaknesses: Array.isArray(aiResult.weaknesses) ? aiResult.weaknesses : [],
    summary: typeof aiResult.summary === 'string' ? aiResult.summary : '',
    createdByAI: true
  });

  await recommendation.save();
  return recommendation;
};

/**
 * Fetch the most recent recruiter recommendation for a candidate.
 */
const getLatestRecruiterRecommendation = async (candidateId) => {
  return RecruiterRecommendation.findOne({ candidateId })
    .sort({ createdAt: -1 })
    .lean();
};

/**
 * Admin: list all recommendations (most recent per candidate).
 */
const getAllRecommendations = async (limit = 50) => {
  return RecruiterRecommendation.find()
    .sort({ createdAt: -1 })
    .limit(limit)
    .populate('candidateId', 'name email')
    .lean();
};

module.exports = {
  generateRecruiterRecommendation,
  getLatestRecruiterRecommendation,
  getAllRecommendations
};

File: backend/services/recruiter_agent_service.py
"""
services/recruiter_agent_service.py — Recruiter agent service (replaces recruiterAgentService.js)
"""
import os
import httpx
import json
import logging
from sqlalchemy.orm import Session
from sqlalchemy import desc
from models.user import User
from models.resume_report import ResumeReport
from models.coding_report import CodingReport
from models.technical_interview_report import TechnicalInterviewReport
from models.hr_interview_report import HRInterviewReport
from models.placement_recommendation import PlacementRecommendation
from models.recruiter_recommendation import RecruiterRecommendation

logger = logging.getLogger(__name__)

def safe_avg(docs: list, field: str) -> float:
    vals = [getattr(d, field) for d in docs if getattr(d, field) is not None]
    nums = [v for v in vals if isinstance(v, (int, float)) and v == v]
    if not nums:
        return None
    return round((sum(nums) / len(nums)) * 10) / 10


def collect_candidate_data(db: Session, user_id: int) -> dict:
    resume_report = (
        db.query(ResumeReport)
        .filter(ResumeReport.user_id == user_id)
        .order_by(desc(ResumeReport.created_at))
        .first()
    )
    coding_reports = (
        db.query(CodingReport)
        .filter(CodingReport.user_id == user_id)
        .order_by(desc(CodingReport.created_at))
        .limit(5)
        .all()
    )
    tech_reports = (
        db.query(TechnicalInterviewReport)
        .filter(TechnicalInterviewReport.user_id == user_id)
        .order_by(desc(TechnicalInterviewReport.created_at))
        .limit(20)
        .all()
    )
    hr_report = (
        db.query(HRInterviewReport)
        .filter(HRInterviewReport.user_id == user_id)
        .order_by(desc(HRInterviewReport.created_at))
        .first()
    )
    placement_report = (
        db.query(PlacementRecommendation)
        .filter(PlacementRecommendation.user_id == user_id)
        .order_by(desc(PlacementRecommendation.created_at))
        .first()
    )

    coding_avg = safe_avg(coding_reports, "score")
    tech_avg = safe_avg(tech_reports, "overall_score")
    tech_comm_avg = safe_avg(tech_reports, "communication_score")

    return {
        "resume": resume_report,
        "codingReports": coding_reports,
        "techReports": tech_reports,
        "hr": hr_report,
        "placement": placement_report,
        "codingAvg": coding_avg,
        "techAvg": tech_avg,
        "techCommAvg": tech_comm_avg,
        "resumeReportId": resume_report.id if resume_report else None,
        "codingReportId": coding_reports[0].id if coding_reports else None,
        "technicalInterviewId": tech_reports[0].id if tech_reports else None,
        "hrInterviewId": hr_report.id if hr_report else None,
        "placementRecommendationId": placement_report.id if placement_report else None,
    }


def build_prompt(data: dict) -> str:
    resume = data["resume"]
    codingAvg = data["codingAvg"]
    techAvg = data["techAvg"]
    techCommAvg = data["techCommAvg"]
    hr = data["hr"]
    placement = data["placement"]

    lines = []

    if resume:
        lines.append(f"ATS Score: {resume.ats_score if resume.ats_score is not None else 'N/A'}/100")
        lines.append(f"Resume Score: {resume.resume_score if resume.resume_score is not None else 'N/A'}/100")
        lines.append(f"Recommended Role: {resume.recommended_role or 'N/A'}")
        lines.append(f"Skills: {', '.join(resume.skills[:10]) if resume.skills else 'N/A'}")
        lines.append(f"Resume Strengths: {'; '.join(resume.strengths[:3]) if resume.strengths else 'N/A'}")
        lines.append(f"Resume Weaknesses: {'; '.join(resume.weaknesses[:3]) if resume.weaknesses else 'N/A'}")

    if codingAvg is not None:
        lines.append(f"Coding Avg Score: {codingAvg}/10")
        verdicts = [r.verdict for r in data["codingReports"] if r.verdict]
        if verdicts:
            lines.append(f"Coding Verdicts: {', '.join(verdicts)}")

    if techAvg is not None:
        lines.append(f"Technical Interview Avg Score: {techAvg}/10")
        lines.append(f"Communication Score (Tech): {techCommAvg if techCommAvg is not None else 'N/A'}/10")
        strengths = [r.strength for r in data["techReports"] if r.strength][:3]
        weaknesses = [r.weakness for r in data["techReports"] if r.weakness][:3]
        if strengths:
            lines.append(f"Technical Strengths: {'; '.join(strengths)}")
        if weaknesses:
            lines.append(f"Technical Weaknesses: {'; '.join(weaknesses)}")

    if hr:
        lines.append(f"HR Interview Overall: {hr.overall_score if hr.overall_score is not None else 'N/A'}/10")
        lines.append(f"HR Communication: {hr.communication_score if hr.communication_score is not None else 'N/A'}/10")
        lines.append(f"HR Confidence: {hr.confidence_score if hr.confidence_score is not None else 'N/A'}/10")
        lines.append(f"HR Professionalism: {hr.professionalism_score if hr.professionalism_score is not None else 'N/A'}/10")
        lines.append(f"HR Recommendation: {hr.recommendation or 'N/A'}")

    if placement:
        lines.append(f"Placement Readiness: {placement.overall_readiness or 'N/A'}")
        lines.append(f"Estimated Placement Chance: {placement.estimated_placement_chance if placement.estimated_placement_chance is not None else 'N/A'}%")
        lines.append(f"Placement Recommended Role: {placement.recommended_role or 'N/A'}")
        if placement.strengths:
            lines.append(f"Placement Strengths: {'; '.join(placement.strengths[:3])}")
        if placement.skills_to_improve:
            lines.append(f"Skills to Improve: {', '.join(placement.skills_to_improve[:3])}")

    candidate_profile = "\n".join(lines)

    return f"""You are a senior technical recruiter making a final hiring decision.

Below is the complete AI-generated assessment of a candidate across all evaluation modules.

--- CANDIDATE ASSESSMENT ---
{candidate_profile or 'No assessment data available yet.'}
--- END ASSESSMENT ---

Based on this data, provide your recruiter recommendation.

CRITICAL: Return ONLY a valid JSON object. No markdown. No explanation. No text before or after.
Start with {{ and end with }}.

{{
  "overallScore": 7.5,
  "recommendation": "Hire",
  "riskLevel": "Low",
  "technicalReadiness": 8.0,
  "communicationReadiness": 7.5,
  "cultureFit": 8.0,
  "strengths": ["Quick learner", "Good Java knowledge", "Team player"],
  "weaknesses": ["Needs DSA practice", "Resume missing projects"],
  "summary": "The candidate has demonstrated strong technical readiness and good communication. Low risk, highly recommended for hire."
}}"""


def parse_groq_response(raw: str) -> dict:
    if not raw or not raw.strip():
        return None

    cleaned = raw.strip()
    if cleaned.startswith("```"):
        cleaned = cleaned.replace("```json", "").replace("```", "").strip()

    try:
        return json.loads(cleaned)
    except Exception:
        pass

    # Extract from first { to last }
    first = raw.find("{")
    last = raw.rfind("}")
    if first != -1 and last > first:
        try:
            return json.loads(raw[first:last+1])
        except Exception:
            pass

    return None


async def generate_recruiter_recommendation(db: Session, candidate_id: int):
    GROQ_API_KEY = os.getenv("GROQ_API_KEY", "")
    if not GROQ_API_KEY:
        raise Exception("GROQ_API_KEY not configured")

    data = collect_candidate_data(db, candidate_id)

    has_any_data = (
        data["resume"] is not None
        or data["codingAvg"] is not None
        or data["techAvg"] is not None
        or data["hr"] is not None
        or data["placement"] is not None
    )

    if not has_any_data:
        raise Exception(
            "No assessment data found for this candidate. "
            "Complete at least one evaluation module first."
        )

    prompt = build_prompt(data)

    async with httpx.AsyncClient(timeout=60.0) as client:
        res = await client.post(
            "https://api.groq.com/openai/v1/chat/completions",
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {GROQ_API_KEY}"
            },
            json={
                "model": "openai/gpt-oss-120b",
                "messages": [
                    {
                        "role": "system",
                        "content": "You are a JSON-only API. Respond with ONLY a valid JSON object. Never include markdown, code fences, or any text outside the JSON object."
                    },
                    {"role": "user", "content": prompt}
                ],
                "temperature": 0.2,
                "max_tokens": 1500
            }
        )

        if res.status_code != 200:
            raise Exception(f"GROQ API error: {res.text}")

        groq_data = res.json()
        raw_text = groq_data["choices"][0]["message"]["content"].strip()
        ai_result = parse_groq_response(raw_text)

        if not ai_result:
            logger.error(f"[RecruiterAgent] JSON parse failed. Raw text: {raw_text[:400]}")
            raise Exception("Unable to parse AI response. Please try again.")

        VALID_RECOMMENDATIONS = ["Strong Hire", "Hire", "Maybe", "Need Another Interview", "Reject"]
        VALID_RISKS = ["Low", "Medium", "High"]

        def clamp(v, lo, hi):
            try:
                val = float(v)
                return min(hi, max(lo, val))
            except Exception:
                return 0.0

        recommendation = RecruiterRecommendation(
            candidate_id=candidate_id,
            overall_score=round(clamp(ai_result.get("overallScore"), 0, 10) * 10) / 10,
            technical_assessment=ai_result.get("technicalAssessment", ""),
            hr_assessment=ai_result.get("hrAssessment", ""),
            ai_summary=ai_result.get("summary", ""),
            recommended_role=ai_result.get("recommendedRole", ""),
            salary_range=ai_result.get("salaryRange", ""),
            hire_recommendation=ai_result.get("recommendation", "Maybe") if ai_result.get("recommendation") in VALID_RECOMMENDATIONS else "Maybe",
            strengths=ai_result.get("strengths", []),
            concerns=ai_result.get("weaknesses", []),
            skills_match=ai_result.get("skillsMatch", []),
            next_steps=ai_result.get("nextSteps", []),
            created_by_ai=True
        )

        db.add(recommendation)
        db.commit()
        db.refresh(recommendation)
        return recommendation


def get_latest_recruiter_recommendation(db: Session, candidate_id: int):
    return (
        db.query(RecruiterRecommendation)
        .filter(RecruiterRecommendation.candidate_id == candidate_id)
        .order_by(desc(RecruiterRecommendation.created_at))
        .first()
    )


def get_all_recommendations(db: Session, limit: int = 50):
    return (
        db.query(RecruiterRecommendation)
        .order_by(desc(RecruiterRecommendation.created_at))
        .limit(limit)
        .all()
    )
